controllers/SnsController.php

```diff
@@ -0,0 +1,115 @@
+<?php
+
+/*
+    ***** BEGIN LICENSE BLOCK *****
+
+    This file is part of the Zotero Data Server.
+
+    Copyright © 2017 Center for History and New Media
+                     George Mason University, Fairfax, Virginia, USA
+                     http://zotero.org
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Affero General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Affero General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+    ***** END LICENSE BLOCK *****
+*/
+
+class SnsController extends Controller
+{
+
+	public function init()
+	{
+
+	}
+
+	public function register($hash, $size)
+	{
+		$results = Zotero_Storage::getUploadQueueItemsAndInfo($hash);
+
+		print_r($results);
+
+		foreach ($results as $result) {
+
+			$info = $result['info'];
+			$item = $result['item'];
+
+			if ($info->size != $size) {
+				error_log("Uploaded file size does not match "
+					. "({$info->size} != {$size}) "
+					. "for file {$info->hash}/{$info->filename}");
+				continue;
+			}
+
+			Zotero_DB::profileStart();
+			Zotero_DB::readOnly(false);
+
+			Zotero_DB::query("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE");
+			Zotero_DB::beginTransaction();
+
+			$fileInfo = Zotero_Storage::getLocalFileInfo($info);
+			if ($fileInfo) {
+				$storageFileID = $fileInfo['storageFileID'];
+			} else {
+				$storageFileID = Zotero_Storage::addFile($info);
+			}
+
+			Zotero_Storage::updateFileItemInfo($item, $storageFileID, $info, true);
+
+			Zotero_Storage::logUpload($info->userID, $item, $info->uploadKey, IPAddress::getIP());
+
+			Zotero_DB::commit();
+		}
+	}
+
+	public function sns()
+	{
+		$username = $_SERVER['PHP_AUTH_USER'];
+		$password = $_SERVER['PHP_AUTH_PW'];
+
+		if ($username != 'user' || $password != 'password') return;
```

Comment:
Expected user and pass should come from values set in config.inc.php, with empty defaults set in config.inc.php-sample.

This should return a 401 if username and password aren't supplied via Basic Auth, 403 for incorrect password, and 500 if username and password aren't set in config.inc.php.